File: backend/app/schemas/user_interest_profiles.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.scoring_profile import LotScoringProfile


NotificationPriorityThreshold = Literal["urgent", "high", "medium", "low"]


class UserInterestProfileBase(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    profile_payload: dict = Field(default_factory=dict)
    min_rating: int = Field(default=0, ge=0, le=100)
    notification_priority_threshold: NotificationPriorityThreshold | None = "medium"
    telegram_enabled: bool = True
    is_active: bool = True

    @field_validator("profile_payload")
    @classmethod
    def validate_profile_payload(cls, value: dict) -> dict:
        return LotScoringProfile.model_validate(value).canonical_payload()


class UserInterestProfileCreate(UserInterestProfileBase):
    pass


class UserInterestProfileUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=160)
    profile_payload: dict | None = None
    min_rating: int | None = Field(default=None, ge=0, le=100)
    notification_priority_threshold: NotificationPriorityThreshold | None = None
    telegram_enabled: bool | None = None
    is_active: bool | None = None

    @field_validator("profile_payload")
    @classmethod
    def validate_profile_payload(cls, value: dict | None) -> dict | None:
        if value is None:
            return None
        return LotScoringProfile.model_validate(value).canonical_payload()


class UserInterestProfileResponse(UserInterestProfileBase):
    id: str
    owner_user_id: str
    created_at: datetime | None = None
    updated_at: datetime | None = None

File: frontend/src/types/userInterestProfiles.ts
export type LotScoringProfilePayload = {
  profile_identifier?: string | null
  target_regions?: string[]
  target_categories?: string[]
  budget_min?: string | number | null
  budget_max?: string | number | null
  minimum_roi?: string | number | null
  minimum_discount?: string | number | null
  allowed_legal_risks?: string[]
  max_distance_km?: string | number | null
  stop_words?: string[]
  desired_keywords?: string[]
  strategy?: 'conservative' | 'balanced' | 'aggressive'
  weights?: Record<string, string | number>
}

export type UserInterestProfile = {
  id: string
  owner_user_id: string
  name: string
  profile_payload: LotScoringProfilePayload
  min_rating: number
  notification_priority_threshold: 'urgent' | 'high' | 'medium' | 'low' | null
  telegram_enabled: boolean
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export type UserInterestProfileCreate = {
  name: string
  profile_payload: LotScoringProfilePayload
  min_rating?: number
  notification_priority_threshold?: 'urgent' | 'high' | 'medium' | 'low' | null
  telegram_enabled?: boolean
  is_active?: boolean
}

export type UserInterestProfileUpdate = Partial<UserInterestProfileCreate>
